How to add shell command with pipe?

The way conform.nvim is set up, it seems slightly more complicated commands do not work.

For example, `shellcheck` [can be used to auto-fix some issues](https://github.com/koalaman/shellcheck/issues/1220#issuecomment-594811243) using this, effectively replacing `shellharden`:

```bash
shellcheck --format=diff my_file.sh | git apply
```

This works fine in the terminal, but the following config does not work:

```lua
formatters = {
	shellcheck = {
		command = "shellcheck",
		args = {"--format=diff", "$FILENAME", "|", "git", "apply" },
		stdin = false,
	},
},
```

I assume this is because in the background, conform.nvim uses `fn.system` with a list of strings, where pipes do not work. However, using `fn.system` with a single string does work fine, but passing a string to `args` results in an error.

Is there a way to make this work?

thx, indeed the change works as it should, that is the using a string as `args`.

Only problem is, the related shell command does not play nice with temp files, it seems? 🤔 

```bash
shellcheck foobar.sh --format=diff
```

```diff
--- a/foobar.sh
+++ b/foobar.sh
@@ -1,1 +1,1 @@
-echo $SHELL
+echo "$SHELL"
```

Reading via stdin does not work here, since the diff is missing the filenames for `git apply`:

```diff
--- a/-
+++ b/-
```

Reading the file directly apparently has the issue that the temp files apparently cause some issue, too. Here the error message from `:ConformInfo`
```
07:38:23[ERROR] Formatter 'shellcheck' error: shellcheck: shellcheck: openBinaryFile: does not exist (No such file or directory)
error: zsh/Users/chrisgrieser/.config/zsh/.conform.6310933.foobar.sh: No such file or directory
```


for reference, here is the config I am using:
```lua
shellcheck = {
	command = "shellcheck",
	args = "shellcheck $FILENAME --shell=bash --format=diff | git apply",
	stdin = false,
},
```

---

In case this is not directly solvable, an alternative approach for this particular case might be for conform.nvim to use the diff resulting from `shellcheck --format=diff` and apply the changes, since it has some diff-applying method anyway?

hmm, using `| patch -p1` also does not seem to work 

```
11:40:15[INFO] Run shellcheck on /Users/chrisgrieser/.config/zsh/foobar.sh
11:40:15[DEBUG] Creating temp file /Users/chrisgrieser/.config/zsh/.conform.6310933.foobar.sh
11:40:15[DEBUG] Run command: shellcheck shellcheck /Users/chrisgrieser/.config/zsh/.conform.6310933.foobar.sh --shell=bash --format=diff | patch -p1
11:40:15[INFO] shellcheck exited with code 1
11:40:15[DEBUG] shellcheck stdout: nil
11:40:15[DEBUG] shellcheck stderr: { "shellcheck: shellcheck: openBinaryFile: does not exist (No such file or directory)", "" }
11:40:15[DEBUG] Cleaning up temp file /Users/chrisgrieser/.config/zsh/.conform.6310933.foobar.sh
11:40:15[ERROR] Formatter 'shellcheck' error: shellcheck: shellcheck: openBinaryFile: does not exist (No such file or directory)
```

--- a/README.md
+++ b/README.md
@@ -260,6 +260,7 @@ require("conform").setup({
       command = "my_cmd",
       -- OPTIONAL - all fields below this are optional
       -- A list of strings, or a function that returns a list of strings
+      -- Return a single string instead to run the command in a shell
       args = { "--stdin-from-filename", "$FILENAME" },
       -- If the formatter supports range formatting, create the range arguments here
       range_args = function(ctx)

--- a/doc/conform.txt
+++ b/doc/conform.txt
@@ -50,6 +50,7 @@ OPTIONS                                                          *conform-option
           command = "my_cmd",
           -- OPTIONAL - all fields below this are optional
           -- A list of strings, or a function that returns a list of strings
+          -- Return a single string instead to run the command in a shell
           args = { "--stdin-from-filename", "$FILENAME" },
           -- If the formatter supports range formatting, create the range arguments here
           range_args = function(ctx)

--- a/lua/conform/init.lua
+++ b/lua/conform/init.lua
@@ -9,8 +9,8 @@ local M = {}
 
 ---@class (exact) conform.FormatterConfig
 ---@field command string|fun(ctx: conform.Context): string
----@field args? string[]|fun(ctx: conform.Context): string[]
----@field range_args? fun(ctx: conform.RangeContext): string[]
+---@field args? string|string[]|fun(ctx: conform.Context): string|string[]
+---@field range_args? fun(ctx: conform.RangeContext): string|string[]
 ---@field cwd? fun(ctx: conform.Context): nil|string
 ---@field require_cwd? boolean When cwd is not found, don't run the formatter (default false)
 ---@field stdin? boolean Send buffer contents to stdin (default true)

--- a/lua/conform/runner.lua
+++ b/lua/conform/runner.lua
@@ -48,12 +48,13 @@ end
 
 ---@param ctx conform.Context
 ---@param config conform.FormatterConfig
+---@return string|string[]
 M.build_cmd = function(ctx, config)
   local command = config.command
   if type(command) == "function" then
     command = command(ctx)
   end
-  local cmd = { command }
+  ---@type string|string[]
   local args = {}
   if ctx.range and config.range_args then
     ---@cast ctx conform.RangeContext
@@ -67,16 +68,22 @@ M.build_cmd = function(ctx, config)
     end
   end
 
-  ---@diagnostic disable-next-line: param-type-mismatch
-  for _, v in ipairs(args) do
-    if v == "$FILENAME" then
-      v = ctx.filename
-    elseif v == "$DIRNAME" then
-      v = ctx.dirname
+  if type(args) == "string" then
+    local interpolated = args:gsub("$FILENAME", ctx.filename):gsub("$DIRNAME", ctx.dirname)
+    return command .. " " .. interpolated
+  else
+    local cmd = { command }
+    ---@diagnostic disable-next-line: param-type-mismatch
+    for _, v in ipairs(args) do
+      if v == "$FILENAME" then
+        v = ctx.filename
+      elseif v == "$DIRNAME" then
+        v = ctx.dirname
+      end
+      table.insert(cmd, v)
     end
-    table.insert(cmd, v)
+    return cmd
   end
-  return cmd
 end
 
 ---@param range? conform.Range

--- a/scripts/options_doc.lua
+++ b/scripts/options_doc.lua
@@ -36,6 +36,7 @@ require("conform").setup({
       command = "my_cmd",
       -- OPTIONAL - all fields below this are optional
       -- A list of strings, or a function that returns a list of strings
+      -- Return a single string instead to run the command in a shell
       args = { "--stdin-from-filename", "$FILENAME" },
       -- If the formatter supports range formatting, create the range arguments here
       range_args = function(ctx)

--- a/tests/runner_spec.lua
+++ b/tests/runner_spec.lua
@@ -102,6 +102,49 @@ describe("runner", function()
       local cmd = runner.build_cmd(ctx, config)
       assert.are.same({ "echo", "--stdin" }, cmd)
     end)
+
+    it("replaces $FILENAME in string args", function()
+      vim.cmd.edit({ args = { "README.md" } })
+      local bufnr = vim.api.nvim_get_current_buf()
+      conform.formatters.test = {
+        meta = { url = "", description = "" },
+        command = "echo",
+        args = "$FILENAME | patch",
+      }
+      local config = assert(conform.get_formatter_config("test"))
+      local ctx = runner.build_context(0, config)
+      local cmd = runner.build_cmd(ctx, config)
+      assert.equal("echo " .. vim.api.nvim_buf_get_name(bufnr) .. " | patch", cmd)
+    end)
+
+    it("replaces $DIRNAME in string args", function()
+      vim.cmd.edit({ args = { "README.md" } })
+      local bufnr = vim.api.nvim_get_current_buf()
+      conform.formatters.test = {
+        meta = { url = "", description = "" },
+        command = "echo",
+        args = "$DIRNAME | patch",
+      }
+      local config = assert(conform.get_formatter_config("test"))
+      local ctx = runner.build_context(0, config)
+      local cmd = runner.build_cmd(ctx, config)
+      assert.equal("echo " .. vim.fs.dirname(vim.api.nvim_buf_get_name(bufnr)) .. " | patch", cmd)
+    end)
+
+    it("resolves arg function with string results", function()
+      vim.cmd.edit({ args = { "README.md" } })
+      conform.formatters.test = {
+        meta = { url = "", description = "" },
+        command = "echo",
+        args = function()
+          return "| patch"
+        end,
+      }
+      local config = assert(conform.get_formatter_config("test"))
+      local ctx = runner.build_context(0, config)
+      local cmd = runner.build_cmd(ctx, config)
+      assert.equal("echo | patch", cmd)
+    end)
   end)
 
   describe("e2e", function()

